text/0101-revamping-the-sdk-performance-api.md

```diff
@@ -0,0 +1,273 @@
+5. Transactions have a max child span count of 1000 which means that eventually data is lost if you keep adding child spans to a transaction.
+
+# Improvements
+
+## New SDK API
+
+The new SDK API has the following requirements:
+
+1. Newly created spans must have the correct trace and parent/child relationship
+2. Users shouldn’t be burdened with knowing if something is a span/transaction
+3. Spans only need a name to identify themselves, everything else is optional.
+4. The new top level APIs should be as similar to the OpenTelemetry SDK public API as possible.
+
+There are two top level methods we'll be introducing to achieve this: `Sentry.startActiveSpan` and `Sentry.startSpan`. `Sentry.startActiveSpan` will take a callback and start/stop a span automatically. In addition, it'll also set the span on the current scope. Under the hood, the SDK will create a transaction or span based on if there is already an existing span on the scope.
+
+```ts
+namespace Sentry {
+  declare function startActiveSpan<T>(
+    spanContext: SpanContext,
+    callback: (span: Span) => T
+  ): T;
+}
+
+// span that is created is provided to callback in case additional
+// attributes have to be added.
+// ideally callback can be async/sync
+const returnVal = Sentry.startActiveSpan(
+  { name: "expensiveCalc" },
+  (span: Span) => expensiveCalc()
+);
+
+// If the SDK needs async/sync typed different we can expose this
+// declare function startActiveSpanAsync<T>(
+//   spanContext: SpanContext,
+//   callback: (span: Span) => Promise<T>,
+// ): Promise<T>;
+```
+
+In the ideal case, `startActiveSpan` should generally follow this code path.
+
+1. Get the active span from the current scope
+2. If the active span is defined, create a child span of that active span based on the provided `spanContext`, otherwise create a transaction based on the provided `spanContext`.
+3. Run the provided callback
+4. Finish the child span/transaction created in step 2
+5. Remove the child span/transaction from the current scope and if it exists, set the previous active span as the active span in the current scope.
+
+If the provided callback throws an exception, the span/transaction created in step 2 should be marked as errored. This error should not be swallowed by `startActiveSpan`.
+
+`startActiveSpan` only provides the correct parent-child relationship if your platform has proper support for forking scopes. For platforms have a single hub/scope (like the mobile SDKs), this method will not lead to the correct parent-child relationship. The SDK will have to provide a different method for these platforms. One option is for `startActiveSpan` to always attach the span it creates to the root span (the transaction), which means users don't get exact parent-child relationships, but they do get relative relationships between spans using relative durations.
+
+`Sentry.startSpan` will create a span, but not set it as the active span in the current scope.
```

Comment:
Any use-cases come to mind why this function is necessary?

Reply:
Mobile has single Hub/Scope and this is needed in case you want your own transaction instead of the auto-instrumented ones, measuring something side by side.
For example, a auto-instrumented transaction during screen loading, but also a service doing the X job, that has nothing to do with the screen loading transaction.